tree/ntuple/v7/test/ntuple_processor_join.cxx

```diff
@@ -0,0 +1,181 @@
+#include "ntuple_test.hxx"
+
+#include <ROOT/RNTupleProcessor.hxx>
+
+TEST(RNTupleJoinProcessor, Basic)
+{
+   FileRaii fileGuard("test_ntuple_join_processor_basic.root");
+   {
+      auto model = RNTupleModel::Create();
+      auto fldX = model->MakeField<float>("x");
+      auto ntuple = RNTupleWriter::Recreate(std::move(model), "ntuple", fileGuard.GetPath());
+
+      for (unsigned i = 0; i < 5; ++i) {
+         *fldX = static_cast<float>(i);
+         ntuple->Fill();
+      }
+   }
+
+   std::vector<RNTupleSourceSpec> ntuples;
+   try {
+      auto proc = RNTupleProcessor::CreateJoin(ntuples);
+      FAIL() << "creating a processor without at least one RNTuple should throw";
+   } catch (const RException &err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("at least one RNTuple must be provided"));
+   }
+
+   ntuples = {{"ntuple", fileGuard.GetPath()}};
+
+   int nEntries = 0;
+   auto proc = RNTupleProcessor::CreateJoin(ntuples);
+   for (const auto &entry : *proc) {
+      auto x = entry.GetPtr<float>("x");
+      EXPECT_FLOAT_EQ(static_cast<float>(proc->GetNEntriesProcessed()), *x);
+      ++nEntries;
+   }
+   EXPECT_EQ(nEntries, 5);
+   EXPECT_EQ(nEntries, proc->GetNEntriesProcessed());
+}
+
+TEST(RNTupleJoinProcessor, Aligned)
+{
+   FileRaii fileGuard1("test_ntuple_join_processor_aligned1.root");
+   {
+      auto model = RNTupleModel::Create();
+      auto fldX = model->MakeField<float>("x");
+      auto ntuple = RNTupleWriter::Recreate(std::move(model), "ntuple1", fileGuard1.GetPath());
+
+      for (unsigned i = 0; i < 5; ++i) {
+         *fldX = static_cast<float>(i);
+         ntuple->Fill();
+      }
+   }
+   FileRaii fileGuard2("test_ntuple_join_processor_aligned2.root");
+   {
+      auto model = RNTupleModel::Create();
+      auto fldY = model->MakeField<std::vector<float>>("y");
+      auto ntuple = RNTupleWriter::Recreate(std::move(model), "ntuple2", fileGuard2.GetPath());
+
+      for (unsigned i = 0; i < 5; ++i) {
+         *fldY = {static_cast<float>(i * 0.2), 3.14, static_cast<float>(i * 1.3)};
+         ntuple->Fill();
+      }
+   }
+   try {
+      std::vector<RNTupleSourceSpec> ntuples = {{"ntuple1", fileGuard1.GetPath()}, {"ntuple1", fileGuard1.GetPath()}};
+      auto proc = RNTupleProcessor::CreateJoin(ntuples);
+      FAIL() << "ntuples with the same name cannot be joined horizontally";
+   } catch (const RException &err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("horizontal joining of RNTuples with the same name is not allowed"));
+   }
+
+   std::vector<RNTupleSourceSpec> ntuples = {{"ntuple1", fileGuard1.GetPath()}, {"ntuple2", fileGuard2.GetPath()}};
+
+   auto proc = RNTupleProcessor::CreateJoin(ntuples);
```

Comment:
The reason why we allow passing just the RNTuple specs, without column names for the join index, is because by default we assume the entries to be aligned and sequential processing, right?

Reply:
Yes, exactly. I am aware this might potentially lead to undesired results due to careless (for lack of a gentler word) programming. I guess a way to address this would be to always require the list of column names, but then allow for empty lists to signal alignment.